### PR TITLE
Fix Bugzilla 15315 - can break immutable with std.algorithm.move

### PR DESCRIPTION
**Pull request by ntrel**

### CODE REVIEW

**Comment by ntrel** [comment]
This is breaking multiple buildkite projects. I think I need to deprecate `move` when T is not assignable. But that's easier said than done, with a decent deprecation message anyway. I think we need a feature for `pragma(deprecated, "custom message");`.

**Comment by dkorpel** [comment]
> I think we need a feature for `pragma(deprecated, "custom message");`.

How's that different from the current `deprecated("custom message")`

**Comment by ntrel** [comment]
@dkorpel *DeprecatedAttribute* attaches to a declaration. Inside a function, it doesn't parse a string message (as a local symbol doesn't need deprecating). A pragma would be a declaration on its own and just fire on declaration semantic. It would avoid duplicating a template function body just to have another symbol declared as deprecated, but with a different constraint. Maybe this is a bit niche though.

I've done the deprecation, but it just prints `T` instead of telling the user what `T` is.

**Comment by ntrel** [comment]
buildkite failing on `std.range` unrelated warnings, all else seems OK

File: std/algorithm/mutation.d [hunk 1]
```diff
@@ -1071,6 +1071,14 @@ Params:
...
 */
 void move(T)(ref T source, ref T target)
+if (__traits(compiles, target = T.init))
+{
+    moveImpl(target, source);
+}
+
+deprecated("Can't move into `target` as `T` can't be assigned")
+void move(T)(ref T source, ref T target)
+if (!__traits(compiles, target = T.init))
```
Comment (thewilsonator) [inline std/algorithm/mutation.d:1081]:
```d
template move(T) 
{
    static if (__traits(compiles, target = T.init))
    void move(ref T source, ref T target) {...}
    else
    deprecated("Can't move into `target` as `~T.stringof~` can't be assigned")
    void move(ref T source, ref T target) {...} 
}
```

Comment (ntrel) [reply std/algorithm/mutation.d]: @thewilsonator Can't use IFTI when there's a `static if`. Also I think it's best not to group the two under a template block.

>     deprecated("Can't move into `target` as `~T.stringof~` can't be assigned")

Now implemented.

**Comment by ntrel** [comment]
Tests now passing, aside from unrelated std.range dscanner buildkite.

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**